File: hellospring/src/main/java/kopher/hellospring/PaymentService.java
package kopher.hellospring;

public class PaymentService {
    public Payment prepare() {
        return new Payment();
    }

    public static void main(String[] args) {
        PaymentService service = new PaymentService();
        Payment payment = service.prepare();
        System.out.println(payment);
    }
}

File: spring-principle/src/main/kotlin/jp/kopher/springprinciple/payment/PaymentService.kt
package jp.kopher.springprinciple.payment

import org.springframework.stereotype.Component
import java.math.BigDecimal
import java.time.LocalDateTime

@Component
class PaymentService(
    private val exRateProvider: ExRateProvider,
) {

    fun prepare(
        orderId: Long,
        currency: String,
        foreignCurrencyAmount: BigDecimal,
    ): Payment {
        val exRate = exRateProvider.getExRate(currency)

        val convertedAmount = foreignCurrencyAmount.multiply(exRate)

        val validUntil = LocalDateTime.now().plusMinutes(30)

        return Payment(
            orderId = orderId,
            currency = currency,
            foreignCurrencyAmount = foreignCurrencyAmount,
            exRate = exRate,
            convertedAmount = convertedAmount,
            validUntil = validUntil,
        )
    }
}